xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py

```diff
@@ -0,0 +1,607 @@
+
+
+@dataclass(frozen=True)
+class IRegion:
+    """An immutable region contains a CFG of immutable blocks. IRegions are contained in operations."""
+
+    blocks: IList[IBlock]
+    """Immutable blocks contained in the IRegion. The first block is the entry block."""
+
+    def __hash__(self) -> int:
+        return hash(id(self))
+
+    def __eq__(self, __o: object) -> bool:
+        return self is __o
+
+    @property
+    def block(self) -> IBlock:
+        """
+        Returns the block of a single-block region.
+        Returns an exception if the region is not single-block.
+        """
+        if len(self.blocks) != 1:
+            raise ValueError(
+                "'block' property of IRegion class is only available "
+                "for single-block regions.")
+        return self.blocks[0]
+
+    @property
+    def ops(self) -> IList[IOperation]:
+        """
+        Get the operations of a single-block region.
+        Returns an exception if the region is not single-block.
+        """
+        if len(self.blocks) != 1:
+            raise ValueError(
+                "'ops' property of IRegion class is only available "
+                "for single-block regions.")
+        return self.block.ops
+
+    def __init__(self, blocks: Sequence[IBlock]):
+        """Creates a new immutable region from a sequence of immutable blocks."""
+
+        object.__setattr__(self, "blocks", IList(blocks))
+        self.blocks.freeze()
+
+    @classmethod
+    def from_mutable(
+        cls,
+        blocks: Sequence[Block],
+        value_map: dict[SSAValue, ISSAValue] | None = None,
+        block_map: dict[Block, IBlock] | None = None,
+    ) -> IRegion:
+        """
+        Creates a new immutable region from a sequence of mutable blocks.
+        The value_map and block_map are used to map already known correspondings
+        of mutable values to immutable values and mutable blocks to immutable blocks.
+        """
+        if value_map is None:
+            value_map = {}
+        if block_map is None:
+            block_map = {}
+
+        # adding dummy block mappings so that ops have a successor to reference
+        # when the actual block is created all successor references will be moved
+        # to the correct block
+        for block in blocks:
+            block_map[block] = IBlock([], [])
+
+        immutable_blocks = [
+            IBlock.from_mutable(block, value_map, block_map)
+            for block in blocks
+        ]
+
+        assert (blocks[0].parent is not None)
+        region = IRegion(immutable_blocks)
+
+        # clean up successor references to blocks for ops inside this region
+        for block, imm_block in zip(blocks, region.blocks):
+            dummy_block = block_map[block]
+            for block in region.blocks:
+                for op in block.ops:
+                    try:
+                        dummy_index = op.successors.index(dummy_block)
+                    except ValueError:
```

Comment:
I went for this solution to avoid the duplicated lookup of:
```
if dummy_block in op.successors:
    dummy_index = op.successors.index(dummy_block)
    ...
```
But not sure about the overhead of throwing and catching the Exception anyway, so I changed it to this now.